Additional routines for existing FP objects for subchannel & thm compatibility

CalfSteven opened the pull request.

CalfSteven committed:
Add v,E to p,T conversion to SodiumSat, Flibe

For Flibem FLiNak and basic Helium:
Add T_from_p_h routines to avoid entropy (p,h) calls in subchannel
Add e_from_p_rho
Add T_from_p_h

refs #22439

CalfSteven committed:
add more ph routine definitions
to Flibe add T_from_p_h, add T_from_p_rho, rm T_from_v_p
refs #22439

CalfSteven committed:
add t_from_p_h s_from_h_p e_from_p_rho to simplefluidproperties
Style fix SimpleFP
refs #22439

GiudGiud commented on modules/fluid_properties/include/userobjects/FlibeFluidProperties.h:
```diff
@@ -38,23 +38,21 @@ class FlibeFluidProperties : public SinglePhaseFluidProperties
                           DualReal & dp_de) const override;
 
+  /**
+   * Temperature and its derivatives from pressure and spesific enthalpy
```
```suggestion
   * Temperature and its derivatives from pressure and specific enthalpy
```

GiudGiud commented on modules/fluid_properties/include/userobjects/FlibeFluidProperties.h:
```diff
@@ -38,23 +38,21 @@ class FlibeFluidProperties : public SinglePhaseFluidProperties
+   * Temperature and its derivatives from pressure and spesific enthalpy
+   *
+   * @param[in] h       specific enthalpy (J/kg)
+   * @param[out] rho       density (kg/m$^3$)
```
```suggestion
   * @param[out] rho    density (kg/m$^3$)
```

GiudGiud commented on modules/fluid_properties/include/userobjects/FlibeFluidProperties.h:
```diff
@@ -80,6 +78,15 @@ class FlibeFluidProperties : public SinglePhaseFluidProperties
                           DualReal & dT_de) const override;
 
+  /**
+   * Temperature and its derivatives from pressure and spesific enthalpy
```
```suggestion
   * Temperature and its derivatives from pressure and specific enthalpy
```

GiudGiud commented on modules/fluid_properties/include/userobjects/FlibeFluidProperties.h:
```diff
@@ -80,6 +78,15 @@ class FlibeFluidProperties : public SinglePhaseFluidProperties
+   *
+   * @param[in] h       specific enthalpy (J/kg)
+   * @param[in] p       pressure (Pa)
+   * @param[out] T       temperature (K)
```
```suggestion
   * @param[out] T      temperature (K)
```

GiudGiud commented on modules/fluid_properties/include/userobjects/FlibeFluidProperties.h:
```diff
@@ -235,6 +241,17 @@ class FlibeFluidProperties : public SinglePhaseFluidProperties
    */
   virtual void e_from_p_T(Real p, Real T, Real & e, Real & de_dp, Real & de_dT) const override;
 
+  /**
+   * Pressure and its derivatives from specific volume and specific internal energy
+   *
+   * @param[in] v        specific volume (m$^3$/kg)
+   * @param[in] e        specific internal energy (J/kg)
+   * @param[out] p       pressure (Pa)
+   * @param[out] dp_dv   derivative of pressure w.r.t. specific volume
+   * @param[out] dp_de   derivative of pressure w.r.t. specific internal energy
+   */
```
```suggestion
  /**
   * Specific energy from pressure and density
   *
   * @param[in] p            pressure (Pa)
   * @param[in] rho        density (kg/m3)
   */
```

GiudGiud commented on modules/fluid_properties/include/userobjects/FlinakFluidProperties.h:
```diff
@@ -80,6 +80,24 @@ class FlinakFluidProperties : public SinglePhaseFluidProperties
+   *
+   * @param[in] h       specific enthalpy (J/kg)
+   * @param[in] p       pressure (Pa)
+   * @param[out] T       temperature (K)
```
```suggestion
   * @param[out] T      temperature (K)
```

GiudGiud commented on modules/fluid_properties/include/userobjects/FlinakFluidProperties.h:
```diff
@@ -80,6 +80,24 @@ class FlinakFluidProperties : public SinglePhaseFluidProperties
                           DualReal & dT_de) const override;
 
+  /**
+   * Temperature and its derivatives from pressure and spesific enthalpy
```
```suggestion
   * Temperature from pressure and specific enthalpy
```

GiudGiud commented on modules/fluid_properties/include/userobjects/FlinakFluidProperties.h:
```diff
@@ -80,6 +80,24 @@ class FlinakFluidProperties : public SinglePhaseFluidProperties
+  virtual Real T_from_p_h(Real p, Real h) const override;
+
+  /**
+   * Temperature and its derivatives from pressure and spesific enthalpy
```
```suggestion
   * Temperature from pressure and density
```

GiudGiud commented on modules/fluid_properties/include/userobjects/FlinakFluidProperties.h:
```diff
@@ -80,6 +80,24 @@ class FlinakFluidProperties : public SinglePhaseFluidProperties
+  /**
+   * Temperature and its derivatives from pressure and spesific enthalpy
+   *
+   * @param[in] h       specific enthalpy (J/kg)
+   * @param[in] p       pressure (Pa)
```
```suggestion
   * @param[in] p       pressure (Pa)
   * @param[in] h       specific enthalpy (J/kg)
```

GiudGiud commented on modules/fluid_properties/include/userobjects/FlinakFluidProperties.h:
```diff
@@ -80,6 +80,24 @@ class FlinakFluidProperties : public SinglePhaseFluidProperties
+   * Temperature and its derivatives from pressure and spesific enthalpy
+   *
+   * @param[in] h       specific enthalpy (J/kg)
+   * @param[out] rho       density (kg/m$^3$)
```
```suggestion
   * @param[out] rho    density (kg/m$^3$)
```

GiudGiud commented on modules/fluid_properties/include/userobjects/FlinakFluidProperties.h:
```diff
@@ -235,6 +253,15 @@ class FlinakFluidProperties : public SinglePhaseFluidProperties
+  /**
+   * Specific internal energy from pressure and density
+   *
+   * @param[in] p   pressure (Pa)
```
```suggestion
   * @param[in] p       pressure (Pa)
```

GiudGiud commented on modules/fluid_properties/include/userobjects/HeliumFluidProperties.h:
```diff
@@ -100,6 +100,15 @@ class HeliumFluidProperties : public SinglePhaseFluidProperties
                   DualReal & dT_de) const override;
 
+  /**
+   * Temperature and its derivatives from pressure and spesific enthalpy
```
```suggestion
   * Temperature from pressure and specific enthalpy
```

GiudGiud commented on modules/fluid_properties/include/userobjects/HeliumFluidProperties.h:
```diff
@@ -100,6 +100,15 @@ class HeliumFluidProperties : public SinglePhaseFluidProperties
+  /**
+   * Temperature and its derivatives from pressure and spesific enthalpy
+   *
+   * @param[in] h       specific enthalpy (J/kg)
+   * @param[in] p       pressure (Pa)
```
```suggestion
   * @param[in] p       pressure (Pa)
   * @param[in] h       specific enthalpy (J/kg)
```

GiudGiud commented on modules/fluid_properties/src/userobjects/SodiumSaturationFluidProperties.C:
```diff
@@ -104,6 +111,22 @@ SodiumSaturationFluidProperties::h_from_p_T(
   dh_dT = cp_from_p_T(pressure, temperature);
 }
 
+DualReal
+SodiumSaturationFluidProperties::P_from_v_e(const DualReal & v,
+                                            const DualReal & e) const ////////////
+{
+  DualReal temperature = T_from_v_e(v, e);
+  // h does not depend on presure
+  return (e - h_from_p_T(0, temperature)) / v;
+}
+
+DualReal
+SodiumSaturationFluidProperties::P_from_v_e(const DualReal & v,
+                                            const DualReal & e) const /////////////
+{
+  return rho_from_p_T;
+}
+
```
```suggestion
```

GiudGiud commented on modules/fluid_properties/src/userobjects/SimpleFluidProperties.C:
```diff
@@ -312,6 +338,26 @@ SimpleFluidProperties::e_from_p_T(
+void
+SimpleFluidProperties::e_from_p_rho(Real p, Real rho, Real & e, Real & de_dp, Real & de_drho) const
+{
+  // get temerature and dirivitives
```
```suggestion
  // get temperature and derivatives
```

GiudGiud commented on modules/fluid_properties/src/userobjects/SimpleFluidProperties.C:
```diff
@@ -312,6 +338,26 @@ SimpleFluidProperties::e_from_p_T(
+  Real T, dT_dp, dT_drho;
+  T_from_p_rho(p, rho, T, dT_dp, dT_drho);
+
+  // get energy and dirivitives
```
```suggestion
  // get energy and derivatives
```

GiudGiud commented on modules/fluid_properties/src/userobjects/SodiumSaturationFluidProperties.C:
```diff
@@ -104,6 +111,22 @@ SodiumSaturationFluidProperties::h_from_p_T(
+                                            const DualReal & e) const ////////////
+{
+  DualReal temperature = T_from_v_e(v, e);
+  // h does not depend on presure
```
```suggestion
  // h does not depend on pressure
```

GiudGiud committed:
Spelling upgrade
Apply suggestions from code review

GiudGiud committed:
Fix mistakes in modifications to SodiumSaturationFluidProps, refs #22439

GiudGiud committed:
Add unit test for sodium saturation FP, refs #22439

GiudGiud committed:
Add unit tests for new routines in Flibe Flinak and Helium FP, refs #22439

GiudGiud committed:
Add unit test for SimpleFluidProps, refs #22439

GiudGiud committed:
Fix mistakes in implementations of p,rho and p,h formulations and derivatives in SimpleFP

GiudGiud committed:
Fix mistakes in implementing T_p_rho in flibe and flinak

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1209351/) on 484f30a wanted to post the following:

The following commit(s) contain fixups:

- 484f30acaf4: fixup! Spelling upgrade Apply suggestions from code review

GiudGiud commented:
@joshuahansel this should be ready for review now

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1209411/) on edee387 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21474/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1209424/) on edee387 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>b92082</th>
      <th colspan=4>#21474 edee38</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.21%</td>
      <td>84.20%</td>
      <td>-0.01%</td>
      <td>-</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>80188</td>
      <td>80183</td>
      <td>-5</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15036</td>
      <td>15041</td>
      <td>+5</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21474/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21474/coverage/framework)


## Modules coverage

### Fluid properties

<table>
  <thead>
    <tr>
      <th></th>
      <th>b92082</th>
      <th colspan=4>#21474 edee38</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>84.58%</td>
      <td>84.67%</td>
      <td>+0.09%</td>
      <td>95.74%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>6104</td>
      <td>6146</td>
      <td>+42</td>
      <td>45</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>1113</td>
      <td>1113</td>
      <td>-</td>
      <td>2</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21474/coverage_diff/fluid_properties)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21474/coverage/fluid_properties)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/richards)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21474/coverage/xfem)

</details>



This comment will be updated on new commits.

joshuahansel commented on modules/fluid_properties/src/userobjects/SodiumSaturationFluidProperties.C:
```diff
@@ -63,17 +63,10 @@ SodiumSaturationFluidProperties::rho_from_p_T(const DualReal & pressure,
-
 Real
-SodiumSaturationFluidProperties::v_from_p_T(Real /* pressure */, Real temperature) const
+SodiumSaturationFluidProperties::v_from_p_T(const Real pressure, const Real temperature) const
```
It's not wrong to pass by constant value, but it is a bit unusual. Since passing by value, the original values are safe. By marking these `const`, you're just ensuring that the local values do not change. Anyway, besides that, these signatures (`T_from_p_T`, `p_from_v_e`, and `T_from_v_e`) do not match the declaration in the header file.

GiudGiud replied:
removed.
smh the different signatures is not getting picked up by compiler.

joshuahansel commented on modules/fluid_properties/unit/src/FlibeFluidPropertiesTest.C:
```diff
@@ -106,6 +106,9 @@ TEST_F(FlibeFluidPropertiesTest, specificInternalEnergy)
   DERIV_TEST(_fp->e_from_p_T, p, T, REL_TOL_DERIVATIVE);
+
+  ABS_TEST(
+      _fp->e_from_p_rho(p, _fp->rho_from_p_T(p, T)), 1.9326997921944547e+06, REL_TOL_SAVED_VALUE);
```
This same value is used above. Just define it once (`const Real e = 1.9326997921944547e+06`)

joshuahansel commented on modules/fluid_properties/unit/src/FlinakFluidPropertiesTest.C:
```diff
@@ -106,6 +106,9 @@ TEST_F(FlinakFluidPropertiesTest, specificInternalEnergy)
   DERIV_TEST(_fp->e_from_p_T, p, T, REL_TOL_DERIVATIVE);
+
+  ABS_TEST(
+      _fp->e_from_p_rho(p, _fp->rho_from_p_T(p, T)), 1607858.289032661588863, REL_TOL_SAVED_VALUE);
```
same

GiudGiud commented on modules/fluid_properties/include/userobjects/FlibeFluidProperties.h:
```diff
@@ -80,6 +78,15 @@ class FlibeFluidProperties : public SinglePhaseFluidProperties
+   * @param[in] h       specific enthalpy (J/kg)
+   * @param[in] p       pressure (Pa)
+   * @param[out] T      temperature (K)
+   */
```
will do

GiudGiud replied:
all these files include the whole docstrings for all routines you sure you want to remove them?

GiudGiud replied:
At this point we paid the price. I d leave them if you re fine with it

joshuahansel replied:
Yes, feel free to leave them

GiudGiud committed:
Address Josh s review:
- use a variable for shared expected test results
- dont pass const value arguments in function defintions in sodiumSat
Refs #22439

joshuahansel commented on modules/fluid_properties/include/userobjects/FlibeFluidProperties.h:
```diff
@@ -38,23 +38,21 @@ class FlibeFluidProperties : public SinglePhaseFluidProperties
+   * @param[out] rho    density (kg/m$^3$)
+   * @param[out] T       temperature (K)
+   */
+  virtual Real T_from_p_rho(Real p, Real rho) const;
```
Ah, duh

GiudGiud commented:
good now?
